File: api/src/main/java/com/caibowen/prma/api/model/ExceptionVO.java
package com.caibowen.prma.api.model;


import javax.annotation.Nullable;
import java.io.Serializable;
import java.util.Arrays;
import java.util.List;

/**
 * @author BowenCai
 * @since 26-10-2014.
 */
public class ExceptionVO implements Serializable {

    private static final long serialVersionUID = 8087093751948611040L;

    public long id;

    public String exceptionName;

    @Nullable public String exceptionMessage;

    /**
     * ordered
     */
    @Nullable public StackTraceElement[] stackTraces;


    public ExceptionVO(){}

    public ExceptionVO(String exceptionName, String exceptionMessage, StackTraceElement[] stackTraces) {

        this.id = calculateID(exceptionName, exceptionMessage, stackTraces);
        this.exceptionName = exceptionName;
        this.exceptionMessage = exceptionMessage;
        this.stackTraces = stackTraces;
    }

    public static final long calculateID(String exceptionName, String exceptionMessage, StackTraceElement[] stackTraces) {

        long expID =  (long) exceptionName.hashCode() << 32;
        if (exceptionMessage!= null)
            expID |= exceptionMessage.hashCode() & 0xFFFFFFFFL;

        if (stackTraces == null || stackTraces.length == 0)
            return expID;

        final long kMul = 0x9ddfea08eb382d69L;
        for (StackTraceElement st : stackTraces) {
            long _s = st.hashCode();
            long _a = ((expID) ^ _s ) * kMul;
            _a ^= _a >> 47;
            expID = (_s ^ _a) * kMul;
            expID ^= expID >> 47;
            expID *= kMul;
        }
        return expID;
    }

    public ExceptionVO(long id, String exceptionName, String exceptionMessage, StackTraceElement[] stackTraces) {
        this.id = id;
        this.exceptionName = exceptionName;
        this.exceptionMessage = exceptionMessage;
        this.stackTraces = stackTraces;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getExceptionName() {
        return exceptionName;
    }

    public void setExceptionName(String exceptionName) {
        this.exceptionName = exceptionName;
    }

    public String getExceptionMessage() {
        return exceptionMessage;
    }

    public void setExceptionMessage(String exceptionMessage) {
        this.exceptionMessage = exceptionMessage;
    }

    public StackTraceElement[] getStackTraces() {
        return stackTraces;
    }

    public void setStackTraces(StackTraceElement[] stackTraces) {
        this.stackTraces = stackTraces;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ExceptionVO)) return false;

        ExceptionVO that = (ExceptionVO) o;

        if (id != that.id) return false;
        if (exceptionMessage != null ? !exceptionMessage.equals(that.exceptionMessage) : that.exceptionMessage != null)
            return false;
        if (!exceptionName.equals(that.exceptionName)) return false;
        if (!Arrays.equals(stackTraces, that.stackTraces)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = (int) (id ^ (id >>> 32));
        result = 31 * result + exceptionName.hashCode();
        result = 31 * result + (exceptionMessage != null ? exceptionMessage.hashCode() : 0);
        result = 31 * result + (stackTraces != null ? Arrays.hashCode(stackTraces) : 0);
        return result;
    }

    @Override
    public String toString() {
        return "com.caibowen.prma.api.model.ExceptionVO{" +
                "id=" + id +
                ", exceptionName='" + exceptionName + '\'' +
                ", exceptionMessage='" + exceptionMessage + '\'' +
                ", stackTraces=" + Arrays.toString(stackTraces) +
                '}';
    }
}

File: api/src/main/scala/com/caibowen/prma/api/model/ExceptionVO.scala
package com.caibowen.prma.api.model
import java.lang.{StringBuilder => JStrBuilder}

/**
 * @author BowenCai
 * @since  02/12/2014.
 */
@SerialVersionUID(8087093751948611040L)
case class ExceptionVO(id: Long,
                       name: String,
                       message: Option[String],
                       stackTraces: Option[Vector[StackTraceElement]]) extends Serializable {

  require(name != null, "Exception name cannot be null")

  def this(eN: String, eMsg: String, sts: Vector[StackTraceElement]) {
    this(Helper.hashCombine(eN, eMsg), eN,
      if (eMsg == null) None else Some(eMsg),
      if(sts == null) None else Some(sts)
    )
  }

  override def hashCode(): Int = {
    var result =  (id ^ (id >>> 32)).toInt
    result = 31 * result + name.hashCode
    result = if (message.isEmpty)
              31 * result
            else 31 * result + message.hashCode
    result = if (stackTraces.isEmpty)
              31 * result
            else 31 * result + stackTraces.hashCode
    result
  }

  def prettyJson(implicit json: JStrBuilder): JStrBuilder = {
    val extra = 100 + (if (stackTraces.isDefined) stackTraces.get.size * 128 else 0)
    json.ensureCapacity(json.capacity + extra)

    json.append("{\r\n  \"id\":").append(id)
      .append(",\r\n  \"name\":\"").append(name).append("\"")
    if (message.isDefined)
      json.append(",\r\n  \"message\":\"")
      Helper.quote(message.get)
        .append('\"')

    if (stackTraces.isDefined && stackTraces.get.size > 0) {
      json.append(",\r\n  \"stackTraces\":[")
      stackTraces.get.foreach(Helper.prettyStackTraceJson(_).append(",\r\n"))
      json.deleteCharAt(json.length - 3)
      json.append("]\r\n")
    }
    json.append('}')
  }

  def appendJson(implicit json: JStrBuilder): JStrBuilder = {
    val extra = 100 + (if (stackTraces.isDefined) stackTraces.get.size * 128 else 0)
    json.ensureCapacity(json.capacity + extra)

    json.append("{\"id\":").append(id)
      .append(",\"name\":\"").append(name).append("\"")
    if (message.isDefined) {
      json.append(",\"message\":\"")
      Helper.quote(message.get)
        .append('\"')
    }
    if (stackTraces.isDefined && stackTraces.get.size > 0) {
      json.append(",\"stackTraces\":[")
      stackTraces.get.foreach(Helper.stackTraceJson(_).append(','))
      json.setCharAt(json.length - 1, ']')
    }
    json.append('}')
  }

  override def toString: String = prettyJson(new JStrBuilder(256)).toString

  override def equals(obj: scala.Any): Boolean = obj match {
    case that: ExceptionVO => {
      if (this eq that)
        return true

      if (id != that.id) return false

      if (!(name == that.name)) return false

      val om = that.message
      if (message.isEmpty) {
        if (om.isDefined) return false
      } else {
        if (om.isEmpty || !(message.get equals  om.get)) return false
      }

      val oss = that.stackTraces
      if (stackTraces.isEmpty) {
        if (oss.isDefined) return false
      } else {
        if (oss.isEmpty || !(stackTraces.get == oss.get)) return false
      }

      true
  }
    case _ => false
  }
}